exoscale/api/compute.py

```diff
@@ -0,0 +1,2088 @@
+            type (InstanceType): the instance type
+            template (InstanceTemplate): the instance template
+            root_disk_size (int): the instance root disk size in GB
+            security_groups ([SecurityGroup]): a list of Security Groups to attach
+                the instance to
+            anti_affinity_groups ([AntiAffinityGroup]): a list of Anti-Affinity Groups
+                to place the instance into
+            private_networks ([PrivateNetwork]): a list of Private Networks to attach
+                the instance to
+            enable_ipv6 (bool): a flag indicating whether to enable IPv6 on the public
+                network interface
+            ssh_key (SSHKey): a SSH Key to deploy on the instance
+            user_data (str): a cloud-init user data configuration
+
+        Returns:
+            Instance: the Compute instance created
+        """
+
+        try:
+            res = self.cs.deployVirtualMachine(
+                name=name,
+                displayname=name,
+                zoneid=zone.id,
+                serviceofferingid=type.id,
+                templateid=template.id,
+                rootdisksize=root_disk_size,
+                securitygroupids=list(i.id for i in security_groups)
+                if security_groups
+                else None,
+                affinitygroupids=list(i.id for i in anti_affinity_groups)
+                if anti_affinity_groups
+                else None,
+                networkids=list(i.id for i in private_networks)
+                if private_networks
+                else None,
+                ip6=enable_ipv6,
+                keypair=ssh_key.name if ssh_key else None,
+                userdata=b64encode(bytes(user_data, encoding="utf-8"))
+                if user_data
+                else None,
+            )
+        except CloudStackApiException as e:
+            raise APIException(e.error["errortext"], e.error)
+
+        return Instance.from_cs(self, res["virtualmachine"])
+
+    def list_instances(
+        self, name=None, ids=None, private_network=None, zone=None, **kwargs
+    ):
+        """
+        List Compute instances.
+
+        Parameters:
+            name (str): an Instance name to restrict results to
+            ids ([str]): a list of Instance ID to restrict results to
+            zone (Zone): a zone to restrict results to
+
+        Yields:
+            Instance: the next Compute instance
+        """
+
+        try:
+            _list = self.cs.listVirtualMachines(
+                fetch_list=True,
+                ids=ids,
+                name=name,
+                zoneid=getattr(zone, "id", None),
+                **kwargs,
+            )
+
+            for i in _list:
+                yield Instance.from_cs(self, i)
+        except CloudStackApiException as e:
+            raise APIException(e.error["errortext"], e.error)
+
+    def get_instance(self, id):
```

Comment:
Realizing now it's undocumented, but `ipaddress` is a valid API filter, and unique across instances. It could be an allowed filtering key.

Reply:
I don't understand what modification you're suggesting here 🤔

Reply:
Ah, sorry :) For example in `get_anti_affinity_group`, you support getting a group by its name or its id because both are unique for a given account.

I meant that `get_instance` could allow getting an instance by its id or its IP address in a similar way.

Reply:
Ah right: added in ﻿﻿a01132c :)